### ISSUE TITLE
ML phase 1: evaluation harness (evaluate.py) with fixed holdout window

### ISSUE THREAD

**Issue by edwinsteele** [edwinsteele/fuel-price-signal#26]
## Context

Phase 1 of the ML signal plan ([PLAN_ml_signal.md](../blob/main/PLAN_ml_signal.md)). Labels landed in #23, features are #24. This is the third Phase 1 deliverable: an evaluation harness so every model variant from phase 2 onwards is scored against the **same held-out window** with the **same metrics**.

Without this, comparing experiments is meaningless — different windows have different difficulty (cycle regularity, shock periods, station coverage), so a 0.61 vs 0.59 log loss across different splits says nothing.

## Scope — what's in

A new module `fuel_signal/evaluate.py` that:

1. Defines the canonical train / val / test date ranges as module-level constants
2. Provides a function to split a labels+features DataFrame into those three subsets by `price_date`
3. Provides scoring functions: log loss, Brier score, marginal-rate baseline
4. Provides a results-log appender that writes a row to `experiments/results.csv`

### Public API sketch

```python
TRAIN_END = \"YYYY-MM-DD\"   # last date in train set
VAL_START = \"YYYY-MM-DD\"   # first date in val set (≥ TRAIN_END + 7 days for label-leakage buffer)
VAL_END   = \"YYYY-MM-DD\"
TEST_START = \"YYYY-MM-DD\"
TEST_END = \"YYYY-MM-DD\"

def split(df: pd.DataFrame) -> tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:
    \"\"\"Returns (train, val, test) by price_date. Drops the buffer windows.\"\"\"

def log_loss(y_true: np.ndarray, y_pred: np.ndarray) -> float: ...
def brier(y_true: np.ndarray, y_pred: np.ndarray) -> float: ...

def baseline_prior(df_train: pd.DataFrame) -> float:
    \"\"\"Returns the marginal positive rate from the training set — the 'do nothing' prediction.\"\"\"

def log_experiment(
    name: str, features: list[str],
    holdout_logloss: float, brier: float,
    notes: str = \"\",
) -> None:
    \"\"\"Appends a row to experiments/results.csv with date + git sha.\"\"\"
```

## Choosing the window — things to consider

This is the design call I want to make myself. Capturing the considerations here so I don't lose them:

**Size**

- The cycle is ~46 days. Test should cover ≥3 cycles (~150 days) for stable metrics, ideally 6 months.
- ~2.5M rows total → 6 months ≈ ~150k rows in test. Plenty of statistical power.

**Recency vs regime stability**

- Most recent data reflects current dynamics — preferred for test.
- But: snapshot data only starts 2026-04-27, so the very recent window is sparser per station.
- Reasonable to put test at e.g. mid-2025 to early-2026, leaving the post-snapshot period as a final sanity check.

**Shock periods**

- COVID 2020, 2026 Middle East war (per CLAUDE.md). Both distort cycle behaviour.
- **Train should include shocks** so the model has seen them.
- **Test probably should not be a shock period** — or if it is, you're scoring shock-robustness, which is a different question. Consider keeping a separate \"shock test set\" later.

**Train/val/test, not just train/test**

- Val for hyperparameter tuning (phase 2+ when LightGBM lands), test only touched at the very end.
- If skipped now, will need to retrofit later — just do it once.

**Buffer between splits (label leakage)**

- Labels look 7 days forward. A row in train with `price_date = 2025-06-30` has its label computed from prices through 2025-07-07.
- If val starts 2025-07-01, the train row's label is computed from val data → leakage.
- **Insert a 7-day gap** (≥ horizon_days) between train_end and val_start, and between val_end and test_start. Drop rows in those buffers.

**Date-grouped, not row-random**

- All rows for one `price_date` go to the same split. (Already implied by date-range splitting; just make sure the implementation enforces it.)

**Don't peek**

- Choose the window once, write it as a constant, never adjust based on results. If you find yourself wanting to nudge the dates after seeing scores, that's the cardinal sin — start over with a frozen window from the beginning.

## Sanity test

The harness has one essential test: scoring the marginal-rate baseline against the test set should produce log loss ≈ 0.693 (or close to whatever `-(p log p + (1−p) log(1−p))` gives for the actual class balance). If this number is off, scoring is broken.

```python
def test_baseline_logloss_matches_entropy():
    df = assemble_training_rows(conn)
    train, val, test = split(df)
    p = baseline_prior(train)
    pred = np.full(len(test), p)
    expected = -(p * np.log(p) + (1-p) * np.log(1-p))  # only exact if test class balance == p
    # tolerance reflects difference between train and test class balance
    assert abs(log_loss(test['label'], pred) - expected) < 0.01
```

## Scope — what's NOT in

- No model code, no sklearn import (other than possibly for `log_loss` / `brier_score_loss` — fine to use)
- No walk-forward CV — single fixed holdout for now; walk-forward is a separate issue if/when needed
- No reliability plots (calibration diagnostics) — phase 2 when there's a model to evaluate
- No realised-spend backtest — that's `backtest.py`, separate

## Files likely affected

- `fuel_signal/evaluate.py` (new)
- `tests/test_evaluate.py` (new)
- `experiments/results.csv` (new — initial row is the prior baseline)
- `experiments/.gitkeep` or similar
- `README.md` — note the eval module exists and what the canonical window is

## Acceptance criteria

- [ ] `fuel_signal/evaluate.py` exists with split/scoring/log_experiment functions
- [ ] Train/val/test date constants chosen and documented (with rationale in a docstring)
- [ ] 7-day buffer between splits enforced
- [ ] Baseline-entropy sanity test passes
- [ ] `experiments/results.csv` initialised with the prior baseline row
- [ ] `uv run ruff check . && uv run pytest -q` passes
- [ ] README updated

## References

- Labels: `fuel_signal/labels.py` (PR #23)
- Features: #24
- Plan: `PLAN_ml_signal.md` Phase 1

**Comment by edwinsteele** [comment]
**Note: first baseline entry for `experiments/results.csv`**

When initialising the results log, the first row should be the marginal-rate baseline scored on the test split — the \"do nothing\" floor that every future experiment is compared against.

Suggested shape:

```csv
date,git_sha,model,features,holdout_logloss,brier,notes
YYYY-MM-DD,<sha>,baseline_prior,none,0.693,0.249,marginal-rate baseline (predict P(label=1)=train_prior for every test row)
```

Compute the actual numbers from `baseline_prior(train)` predicted across the test set, don't hardcode 0.693 — they'll differ slightly from the entropy formula because train and test class balances aren't identical.

**Comment by edwinsteele** [comment]
These are shocks to consider. CLAUDE.md may have others that we should review for consideration

1. ~Jan 2019 (~115c/L)
The Q4 2018 global oil price crash. Brent fell ~40% between October and December 2018, driven by record US shale output, Trump administration granting Iran sanctions waivers to eight major buyers (removing the feared supply crunch), and deteriorating global growth expectations. The lowest Sydney prices hit with a ~6-8 week lag from the crude move.
2. ~Mar–Apr 2020 (~90–95c/L)
The COVID crash — the deepest trough on the chart and the only time Sydney prices fell below 100c/L in this period. Two simultaneous shocks: global demand collapsed almost overnight as lockdowns hit, while Saudi Arabia and Russia simultaneously launched a price war after OPEC+ talks broke down in early March. Brent briefly touched ~$20/bbl; WTI went negative. This is the chart's absolute floor.
3. ~Sep 2022 (~155c/L)
As I described above — the federal excise halving (22c/L cut, active 30 March to 29 September 2022) coinciding with crude retreating from its post-Ukraine invasion peak. The sharp snap-back in October is the excise restoration.

There's also a macro event that we may or may not care to think about
1. ~Jul–Aug 2024 (t~145c/L)
Softer crude driven primarily by persistent Chinese demand disappointment — the post-COVID recovery in China never matched forecasts, and by mid-2024 that weakness was clearly structural rather than temporary. Brent drifted from ~$87 to ~$75/bbl around that window. No policy change on the Australian side; this was purely a crude pass-through.
